### PR TITLE
Unittests for esphome python code

### PR DESCRIPTION
**Pull request by timsavage**
## Description:
Configuration and setup of pytest with coverage reporting and hypothesis. Setup of scripts following the established patterns to run the tests. Initial spike has improved coverage to ~16%.

The initial tests have already found (potential) bugs, edge cases. 

Components have been excluded from test cases in the first round.

**Related issue (if applicable):** fixes lack of unit tests (as mentioned in tests/README.md)

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by timsavage** [comment]
One issue I'm coming across is the inconsistent implementation of comparison operators, the main behaviours when an unsupported type is supplied:

- Raise a `ValueError`
- Return `False`

The recommended way is to return `NotImplemented` and allow the Python interpreter to handle the outcome this allows for the inverse operators to be utilised. If a type cannot be compared a `TypeError` is raised.

I can write the test either way, but it would be good for them to be consistent.

**Comment by OttoWinter** [comment]
Wow, this is great, thank you very much!

I haven't used the hypothesis python library before (only heard of it on HN), but it looks really interesting for unit testing.

This is already looking great for an initial setup!

As for the core code (relating to `ValueError`) - I think the `TimePeriod` class for example is still basically the same form when I initially wrote ESPHome(yaml) as a quick experiment (and I didn't put much thought into what would be the correct pythonic behavior, nor did I know those well at the time).

So basically, as long as it works I'm happy with any changes to those classes/functions. I'd also be happy to merge right away if you want to move that to another PR (gets easier to review then).

**Comment by timsavage** [comment]
I was asking questions in the Discord dev channel, there are currently a number of failures, and I was interested in how you wanted to approach these. I can mark them as "expected failures" for now so they don't block things but are obvious they need fixing, or just leave them as failing so they are a bit more obvious they need to be fixed or at least reviewed.

I was aiming to get to 20% coverage before moving to ready for review. I can then work on more tests in parallel. If you are happy with that approach I'm close to the 20% goal.

**Comment by timsavage** [comment]
While far from complete this is the first 20% of the core python code tested. There are still a number of failing tests. 

    ----------- coverage: platform linux, python 3.7.5-final-0 -----------
    Name                             Stmts   Miss Branch BrPart  Cover
    ------------------------------------------------------------------
    esphome/__init__.py                  0      0      0      0   100%
    esphome/__main__.py                371    371    113      0     0%
    esphome/api/__init__.py              0      0      0      0   100%
    esphome/api/api_pb2.py             215    215      2      0     0%
    esphome/api/client.py              353    353    112      0     0%
    esphome/automation.py              168    168     36      0     0%
    esphome/codegen.py                   3      3      0      0     0%
    esphome/config.py                  599    599    308      0     0%
    esphome/config_helpers.py           11     11      2      0     0%
    esphome/config_validation.py       764    537    350      5    24%
    esphome/const.py                   573      0      0      0   100%
    esphome/core.py                    535    178    182      5    65%
    esphome/core_config.py             135    135     58      0     0%
    esphome/cpp_generator.py           433    311    144      0    21%
    esphome/cpp_helpers.py              47     47     18      0     0%
    esphome/cpp_types.py                31     31      0      0     0%
    esphome/dashboard/__init__.py        0      0      0      0   100%
    esphome/dashboard/dashboard.py     504    504    136      0     0%
    esphome/espota2.py                 211    211     56      0     0%
    esphome/helpers.py                 196     73     68      5    63%
    esphome/legacy.py                    8      8      2      0     0%
    esphome/mqtt.py                    106    106     28      0     0%
    esphome/pins.py                    167     47     92      3    68%
    esphome/platformio_api.py          137    137     42      0     0%
    esphome/py_compat.py                 0      0      0      0   100%
    esphome/storage_json.py            107    107      0      0     0%
    esphome/symlink_ops.py             102    102     22      0     0%
    esphome/util.py                    162    128     46      0    16%
    esphome/voluptuous_schema.py       127     76     72     11    37%
    esphome/vscode.py                   45     45      6      0     0%
    esphome/wizard.py                  183    183     24      0     0%
    esphome/writer.py                  227    227     96      0     0%
    esphome/yaml_util.py               255    255     88      0     0%
    esphome/zeroconf.py                510    510    129      0     0%
    ------------------------------------------------------------------
    TOTAL                             7285   5678   2232     29    20%

    6 failed, 305 passed, 1 xfailed in 7.31s

**Comment by brandond** [comment]
LGTM. Awesome work!

**Comment by brandond** [comment]
Lets do it.
